sentry-android-core/src/main/java/io/sentry/android/core/AndroidTraceTransactionListener.java

```diff
@@ -0,0 +1,171 @@
+              "Profiling is already active and was started by transaction %s",
+              activeTransaction.getSpanContext().getTraceId().toString());
+      return;
+    }
+
+    traceFile = new File(traceFilesDir, UUID.randomUUID() + ".trace");
+
+    if (traceFile.exists()) {
+      options
+          .getLogger()
+          .log(SentryLevel.DEBUG, "Trace file already exists: %s", traceFile.getPath());
+      return;
+    }
+    activeTransaction = transaction;
+
+    // We stop the trace after 30 seconds, since such a long trace is very probably a trace
+    // that will never end due to an error
+    scheduledFinish =
+        options
+            .getExecutorService()
+            .schedule(() -> lastProfilingData = onTransactionFinish(transaction), 30_000);
+
+    Debug.startMethodTracingSampling(traceFile.getPath(), BUFFER_SIZE_BYTES, intervalUs);
+  }
+
+  @Override
+  public synchronized @Nullable ProfilingTraceData onTransactionFinish(
+      @NotNull ITransaction transaction) {
+
+    final ITransaction finalActiveTransaction = activeTransaction;
+    // Profiling finished, but we check if we cached last profiling data due to a timeout
+    if (finalActiveTransaction == null) {
+      ProfilingTraceData profilingData = lastProfilingData;
+      // If the cached last profiling data refers to the transaction that started it we return it
+      // back, otherwise we would simply lose it
+      if (profilingData != null
+          && profilingData
+              .getTraceId()
+              .equals(transaction.getSpanContext().getTraceId().toString())) {
+        return profilingData;
+      }
+      return null;
+    }
+
+    if (finalActiveTransaction != transaction) {
+      options
+          .getLogger()
+          .log(
+              SentryLevel.DEBUG,
+              "Transaction %s finished, but profiling was started by transaction %s. Skipping",
+              transaction.getSpanContext().getTraceId().toString(),
+              finalActiveTransaction.getSpanContext().getTraceId().toString());
+      return null;
+    }
+
+    Debug.stopMethodTracing();
+
+    lastProfilingData = null;
+    activeTransaction = null;
+
+    if (scheduledFinish != null) {
+      scheduledFinish.cancel(true);
+    }
+
+    if (traceFile == null || !traceFile.exists()) {
+      options
+          .getLogger()
+          .log(
+              SentryLevel.DEBUG,
+              "Trace file %s does not exists",
+              traceFile == null ? "null" : traceFile.getPath());
+      return null;
+    }
+
+    // todo how to retrieve these fields? (version name and code)
+    //  buildNumber = Build.VERSION.INCREMENTAL? - osName = "android"?
+    return new ProfilingTraceData(
+        traceFile,
+        transaction,
+        Build.VERSION.SDK_INT,
+        Build.MANUFACTURER,
+        Build.MODEL,
+        "buildNumber", // ???
+        "android", // ???
+        Build.VERSION.RELEASE,
+        "versionName", // ???
+        "versionCode", // ???
```

Comment:
We have https://github.com/getsentry/sentry-java/blob/380ea81a7e8b37d3f033e5b7b97885d06fd96fe6/sentry-android-core/src/main/java/io/sentry/android/core/ContextUtils.java#L37 for this, I think you could do the same for `versionName` easily.

Reply:
Yeah, i've seen it, but it requires a `PackageInfo`, which requires a `Context`, and i don't have it anywhere.
Would it be a good idea to have these kind of fields (which require a `PackageInfo`) initialized in `AndroidOptionsInitializer.init()` and save them in the `SentryAndroidOptions`?

Reply:
We can make a sort of abstraction, either on `SentryAndroidOptions` or this class takes the `PackageInfo` via ctor or something, otherwise it cannot be tested it.

Reply:
I ended up adding the `PackageInfo` via ctor